### PR TITLE
feat(dreamzero): enable gradient checkpointing with use_reentrant support

### PR DESCRIPTION
**Pull request by thereAreDemonsNearby**
### Description

1. Add use_reentrant=True path for gradient checkpointing in CausalWanModel to work around CUDA graph compatibility issues. 
2. Wire gradient_checkpointing and gradient_checkpointing_use_reentrant config through FSDP model manager. 
3. Also fix FSDP2 fully_shard wrapping order (inside-out via reversed) and add CUDA graph step marking in sft_forward.

### Motivation and Context
1. Training 5B model without gradient checkpointing limits micro batch size, which harms throughput.
2. `use_reentrant=False` (the DreamZero default) conflicts with CUDA Graph capture, causing a known PyTorch issue (https://github.com/pytorch/pytorch/issues/154306). So we supported the use_reentrant=True code path.
3. There is also a subtle FSDP2 bug that requires wrapping CausalWanModel.
4. Using checkpoint(use_reentrant=False) with FSDP2 causes gpu memory leak in backward passes if PyTorch version is under 2.10 (https://github.com/pytorch/pytorch/issues/169349). So if gradient checkpoint is needed, please use **PyTorch >= 2.10**.

### How has this been tested?

1. Now we can enable gradient checkpointing under the conditions of micro batch size per gpu > 1 and cuda graph enabled, with no error reported,  using PyTorch == 2.10
2. With mbs == 8, 3.6 samples/sec/gpu is achieved, compared to 1.7 samples/sec/gpu under mbs=2
3. However, the current bottleneck in training speed has become data processing. The next PR will fix this.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/dreamzero/dreamzero_policy.py [hunk 1]
```diff
@@ -87,16 +90,31 @@ def __init__(
...
+
             if diffusion_model is not None:
                 if hasattr(diffusion_model, "_set_gradient_checkpointing"):
                     diffusion_model._set_gradient_checkpointing(
                         diffusion_model, enabled
                     )
                 elif hasattr(diffusion_model, "gradient_checkpointing"):
                     diffusion_model.gradient_checkpointing = enabled
+
+            setattr(diffusion_model, 'gradient_checkpointing_use_reentrant', use_reentrant)
```
Comment (qurakchin) [inline rlinf/models/embodiment/dreamzero/dreamzero_policy.py]: have you tested when diffusion_model is None?

File: rlinf/models/embodiment/dreamzero/dreamzero_policy.py [hunk 2]
```diff
@@ -75,10 +75,13 @@ def __init__(self, **kwargs):
 class DreamZeroPolicy(VLA, BasePolicy):
     """Lightweight DreamZero action model: IdentityBackbone + WANPolicyHead."""
 
+    # CausalWanModel has to be wrapped to avoid a FSDP2 bug
+    # when using with gradient checkpointing
     _no_split_modules = [
         "T5SelfAttention",  # text encoder
         "AttentionBlock",  # vae
-        "CausalWanAttentionBlock",  # action head
+        "CausalWanModel", # action head
```
Comment (qurakchin) [inline rlinf/models/embodiment/dreamzero/dreamzero_policy.py]: will this harm the performance?

File: rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py [hunk 3]
```diff
@@ -136,16 +136,36 @@ def custom_forward(*inputs, **kwargs):
...
-            x = torch.utils.checkpoint.checkpoint(
-                create_custom_forward(block),
-                x,
-                **kwargs,
-                use_reentrant=False,
-            )
+
+            ckpt_use_reentrant = getattr(self, 'gradient_checkpointing_use_reentrant', True)
+
+            if not ckpt_use_reentrant:
```
Comment (FxxxxU) [inline rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py]: It would be better to change this to the if ckpt_use_reentrant: style to make it easier to understand.

File: rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py [hunk 4]
```diff
@@ -136,16 +136,36 @@ def custom_forward(*inputs, **kwargs):
...
+                x = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(block),
+                    x,
+                    **kwargs,
+                    use_reentrant=False,
+                )
+            else:
+                # When use_reentrant=True, the checkpoint interface only allows positional arguments
```
Comment (FxxxxU) [inline rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py]: Complete the comment for use_reentrant to gradient_checkpointing_use_reentrant.

**Review by FxxxxU — APPROVED** [review]

File: rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py (hunk 3 above)
Comment (thereAreDemonsNearby) [reply rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py]: done

File: rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py (hunk 4 above)
Comment (thereAreDemonsNearby) [reply rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py]: done

File: rlinf/models/embodiment/dreamzero/dreamzero_policy.py (hunk 1 above)
Comment (thereAreDemonsNearby) [reply rlinf/models/embodiment/dreamzero/dreamzero_policy.py]: In DreamZero, diffusion model can't be None because it's the most important backbone of the network

File: rlinf/models/embodiment/dreamzero/dreamzero_policy.py (hunk 2 above)
Comment (thereAreDemonsNearby) [reply rlinf/models/embodiment/dreamzero/dreamzero_policy.py]: It will not. It just move the all gather of some embeddings from the beginning of the forward pass to the beginning of CausalWanModel forward. And I see no obvious performance degradation.

File: rlinf/models/embodiment/dreamzero/dreamzero_policy.py [hunk 5]
```diff
@@ -87,16 +90,31 @@ def __init__(
...
+
+    # This method is called in FSDPModelManager.setup_model_and_optimizer
+    def gradient_checkpointing_enable(self, gradient_checkpointing_kwargs={}):
         try:
             diffusion_model = getattr(getattr(self, "action_head", None), "model", None)
-            enabled = self.config.gradient_checkpointing
+            enabled = True
+            use_reentrant = gradient_checkpointing_kwargs.get("use_reentrant", True)
+
             if diffusion_model is not None:
```
Comment (qurakchin) [inline rlinf/models/embodiment/dreamzero/dreamzero_policy.py]: should this branch exist if diffusion_model will not be None

Comment (thereAreDemonsNearby) [reply rlinf/models/embodiment/dreamzero/dreamzero_policy.py]: You are right, I'll remove the if and add a check for this

**Review by qurakchin — APPROVED** [review]
